src/app/InteractionModelEngine.cpp

```diff
@@ -751,7 +781,73 @@ bool InteractionModelEngine::EnsureResourceForSubscription(FabricIndex aFabricIn
+bool InteractionModelEngine::TrimFabricForRead(FabricIndex aFabricIndex)
+{
+    constexpr size_t minSupportedPathsPerFabricForRead = kMinSupportedReadRequestsPerFabric * kMinSupportedPathsPerReadRequest;
+
+    size_t attributePathsUsedByCurrentFabric = 0;
+    size_t eventPathsUsedByCurrentFabric     = 0;
+    size_t readTransactionsOnCurrentFabric   = 0;
+
+    ReadHandler * candidate            = nullptr;
+    size_t candicateAttributePathsUsed = 0;
+    size_t candicateEventPathsUsed     = 0;
+
+    // It is safe to use & here since this function will be called on current stack.
+    mReadHandlers.ForEachActiveObject([&](ReadHandler * handler) {
+        if (handler->GetAccessingFabricIndex() != aFabricIndex || !handler->IsType(ReadHandler::InteractionType::Read))
+        {
+            return Loop::Continue;
+        }
+
+        size_t attributePathsUsed = handler->GetAttributePathCount();
+        size_t eventPathsUsed     = handler->GetEventPathCount();
+
+        attributePathsUsedByCurrentFabric += attributePathsUsed;
+        eventPathsUsedByCurrentFabric += eventPathsUsed;
+        readTransactionsOnCurrentFabric++;
+
+        if (candidate == nullptr)
+        {
+            candidate = handler;
+        }
+        // Oversized read handlers will be evicted first.
+        else if ((attributePathsUsed > kMinSupportedPathsPerReadRequest || eventPathsUsed > kMinSupportedPathsPerReadRequest) &&
+                 (candicateAttributePathsUsed <= kMinSupportedPathsPerReadRequest &&
+                  candicateEventPathsUsed <= kMinSupportedPathsPerReadRequest))
+        {
+            candidate                   = handler;
+            candicateAttributePathsUsed = attributePathsUsed;
+            candicateEventPathsUsed     = eventPathsUsed;
+        }
+        // This handler is older than the one we picked before.
+        else if (handler->GetTransactionStartGeneration() < candidate->GetTransactionStartGeneration() &&
+                 // And the level of resource usage is the same (both exceed or neither exceed)
+                 ((attributePathsUsed > kMinSupportedPathsPerReadRequest || eventPathsUsed > kMinSupportedPathsPerReadRequest) ==
+                  (candicateAttributePathsUsed > kMinSupportedPathsPerReadRequest ||
+                   candicateEventPathsUsed > kMinSupportedPathsPerReadRequest)))
+        {
+            candidate = handler;
+        }
+        return Loop::Continue;
+    });
+
+    if (candidate != nullptr &&
+        ((attributePathsUsedByCurrentFabric > minSupportedPathsPerFabricForRead ||
+          eventPathsUsedByCurrentFabric > minSupportedPathsPerFabricForRead ||
+          readTransactionsOnCurrentFabric > kMinSupportedReadRequestsPerFabric) ||
+         // Always evict the transactions on PASE sessions if the fabric table is full.
+         (aFabricIndex == kUndefinedFabricIndex && mpFabricTable->FabricCount() == GetConfigMaxFabrics())))
+    {
+        candidate->Abort();
```

Comment:
No, this makes no sense to me.

If our existing reads were all not oversize, why would we ever evict spec-guaranteed in-progress reads instead of just returning BUSY for the incoming read and having it retry?

And if our caller guarantees we are only called here for oversize reads, then shouldn't we be looking for them in the list, and not even considering as candidates reads that are not oversize?

Reply:
What if the fabric is oversized? As the spec guaranteed the # of reads x # of paths in the read. We should be able to evict some reads if the fabric is oversized.

Reply:
Ah, I see.  Yes, if a fabric is oversized and we are trying to service a _different_ fabric that is not oversized even with a new read, then it makes sense to evict the oversized bit.

But let's say we have enough read handlers for two reads per fabric and each fabric has two reads already and a new read comes in on some fabric.  What should we do in that case?

Reply:
If "some" fabric is one of two, then it will be rejected with busy (test case 3, https://github.com/project-chip/connectedhomeip/pull/18884/files#diff-09d4e6f8f568b8ae1f60c465f4f2c89e3fb4e92d92ac2308fae80036b2430a1dR933)

If "some" fabric is another fabric, then it will be accepted by evicting one on the other fabric. (test case 4, https://github.com/project-chip/connectedhomeip/pull/18884/files#diff-09d4e6f8f568b8ae1f60c465f4f2c89e3fb4e92d92ac2308fae80036b2430a1dR957)

Reply:
I see, your point if what if the resource provided is enough for providing more resources required by the spec.

I think I can update the minimal resource limit to guarantee at least floor(`CHIP_IM_MAX_NUM_READS`/`CHIP_CONFIG_MAX_FABRICS`)

Reply:
Added `GetGuaranteedReadRequestsPerFabric`, the existing test case (2, 3 and 4) should already covered it (since I just replace the constants in the code by GetGuaranteedReadRequestsPerFabric()